File: baekJoon/1780.cpp
#include <iostream>
#include <stdlib.h>

using namespace std;
int ans[3];

bool isTrue(int** arr, int x, int y, int n){
    int a = arr[x][y];
    for(int i=x ; i<x+n; i++)
        for (int j=y; j<y+n; j++)
            if(arr[i][j] != a)
                return false;
    ans[a+1] += 1;
    return true;
}

void update(int** arr, int x, int y, int n){
    if(isTrue(arr,x,y,n))
        return;
    int s = int(n/3);
    for(int i=0; i<3; i++)
        for(int j=0; j<3; j++)
            update(arr, x+i*s, y+j*s, s);
    return;
}

int main(){
    long int n;
    int** arr;
    cin >> n;
    
    arr = (int**)malloc(sizeof(int*)*n);
    for (int i=0; i<n; i++)
        arr[i] = (int*)malloc(sizeof(int)*n);   
    
    for(int i=0; i<n;i++)
        for(int j=0; j<n; j++)
            cin >> arr[i][j];
    update(arr, 0,0,n);

    for(int i=0; i<3; i++)
        cout << ans[i] << endl;
    
    return 0;
}

File: 1780.py
from sys import stdin
global ans
ans = [0,0,0]

def isTrue(Arr, ind, s):
    global ans
    Ans = Arr[ind[0]][ind[1]]
    for i in range(ind[0],ind[0]+s):
        for j in range(ind[1], ind[1]+s):
            if Arr[i][j] != Ans:
                return False
    ans[Ans+1] += 1
    return True

def printArr(Arr):
    for a in Arr:
        print(a)

def copyArr(arr, ind, size):
    tmp = []
    for i in range(size):
        tmp.append([])
        for j in range(size):
            tmp[i].append(arr[ind[0]+i][ind[1]+j])
    return tmp

def update(Arr, ind, size):
    
    if isTrue(Arr, ind, size):
        return
    s = len(Arr) // 3
    for i in range(3):
        for j in range(3):
            update(Arr, [ind[0]+ i*s, ind[1]+j*s], s)
    
    
arr = []
arrSize = int(stdin.readline())
for i in range(arrSize):
    arr.append(list(map(int, stdin.readline().split())))

update(arr, [0,0], len(arr))
for a in ans:
    print(a)